src/main/ank/1.6-simple-algebraic-data-types.md

```diff
@@ -0,0 +1,328 @@
+```
+................
+```Haskell
+stmt = (,) "This statement is" False
+```
+```kotlin:ank:silent
+fun <A, B> `(,)`(fst:A, snd:B): Pair<A, B> = fst to snd
+val stmt = `(,)`("This statement is" , false)
+```
+................
+```Haskell
+data Stmt = Stmt String Bool
+```
+```kotlin:ank:silent
+typealias Stmt = Pair<String, Boolean>
+```
+................
+```Haskell
+startsWithSymbol :: (String, String, Int) -> Bool 
+startsWithSymbol (name, symbol, _) = isPrefixOf symbol name
+```
+```kotlin:ank:silent
+import arrow.core.Tuple3
+
+val startsWithSymbol: (Tuple3<String, String, Int>) -> Boolean =
+    {it.a.startsWith(it.b)}
+```
+................
+```Haskell
+data Element = Element { name :: String
+                       , symbol :: String
+                       , atomicNumber :: Int }
+```
+```kotlin:ank:silent
+data class Element(val name:String, val symbol:String, val atomicNumber:Int)
+```
+................
+```Haskell
+tupleToElem :: (String, String, Int) -> Element
+tupleToElem (n, s, a) = Element { name = n
+                                , symbol = s
+                                , atomicNumber = a }
+```
+```kotlin:ank:silent
+import arrow.core.Tuple3
+
+val tupleToElm: (Tuple3<String, String, Int>) -> Element = 
+    {Element(it.a, it.b, it.c)}
+```
+................
+```Haskell
+elemToTuple :: Element -> (String, String, Int) 
+elemToTuple e = (name e, symbol e, atomicNumber e)
+```
+```kotlin:ank:silent
+import arrow.core.Tuple3
+
+val elemToTuple: (Element) -> Tuple3<String, String, Int> =
+    {Tuple3(it.name, it.symbol, it.atomicNumber)}
+```
+................
+```Haskell
+atomicNumber :: Element -> Int
+```
+```kotlin
+val atomicNumber : (Element) -> Int
+```
+................
+```Haskell
+startsWithSymbol :: Element -> Bool 
+startsWithSymbol e = isPrefixOf (symbol e) (name e)
+```
+```kotlin:ank:silent
+val isPrefixOf = {prefix: String, str:String -> str.startsWith(prefix)}
+
+val startsWithSymbol: (Element) -> Boolean = {isPrefixOf(it.symbol, it.name)} 
+```
+................
+```Haskell
+startsWithSymbol e = symbol e `isPrefixOf` name e
+```
+```kotlin:ank:silent
+infix fun String.isPrefixOf(el:String) = isPrefixOf(this, el)
+
+val startsWithSymbol: (Element) -> Boolean = {it.symbol isPrefixOf it.name}
+```
+................
+```Haskell
+data Either a b = Left a | Right b
+```
+```kotlin:ank:silent
+import arrow.generic.coproduct2.Coproduct2
+import arrow.core.Either.Left
+import arrow.core.Either.Right
+
+typealias Either<A, B> = Coproduct2<Left<A>, Right<B>>
+```
+................
+```Haskell
+data OneOfThree a b c = Sinistral a | Medial b | Dextral c
+```
+```kotlin:ank:silent
+import arrow.generic.coproduct3.Coproduct3
+
+typealias OneOfThree<A, B, C> = Coproduct3<A, B, C>
+```
+................
+```Haskell
+Either a Void
+```
+```kotlin
+Either<A, Nothing>
+```
+................
+```Haskell
+data Color = Red | Green | Blue
+```
+```kotlin
+typealias Color = Coproduct3<Red, Green, Blue>
```

Comment:
This is better expressed with sealed classes and not the coproduct data type which are the equivalent to sum the. Coproduct itself let's you represent that but the Kotlin way to represent this is sealed class adts